For #25381 - Add a setting allowing to toggle Pocket sponsored stories

Mugurell opened the pull request:
Issue #25381 
More details and a video in the tracking Jira ticket - https://mozilla-hub.atlassian.net/browse/FNXV2-20135
Needs https://github.com/mozilla-mobile/android-components/pull/12259 which changes `deleteProfile` to not be a suspending function anymore.

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/utils/Settings.kt:
```diff
@@ -1268,12 +1268,24 @@ class Settings(private val appContext: Context) : PreferencesHolder {
     )
 
+    /**
+     * Indicates if the Pocket recommended stories homescreen section should be shown or not.
```
```suggestion
     * Indicates if the Pocket recommended stories homescreen section should be shown.
```

MozillaNoah commented on app/src/main/res/values/strings.xml:
```diff
@@ -411,6 +411,8 @@
     <string name="customize_toggle_recently_visited">Recently visited</string>
     <!-- Title for the customize home screen section with Pocket. -->
     <string name="customize_toggle_pocket">Pocket</string>
+    <!-- Title for the customization of Pocket home screen section with also containing sponsored stories. -->
```
```suggestion
    <!-- Title for the customize home screen section with sponsored Pocket stories . -->
```

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/settings/HomeSettingsFragment.kt:
```diff
@@ -116,6 +126,44 @@ class HomeSettingsFragment : PreferenceFragmentCompat() {
+                                val wasProfileDeletionSuccessful = withContext(ioDispatcher) {
+                                    context.components.core.pocketStoriesService.deleteProfile()
+                                }
+                                if (wasProfileDeletionSuccessful) {
```
nit: add a new line here

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/components/appstate/AppStoreReducer.kt:
```diff
@@ -148,9 +148,15 @@ internal object AppStoreReducer {
-        is AppAction.PocketSponsoredStoriesChange -> state.copy(
-            pocketSponsoredStories = action.sponsoredStories
-        )
+        is AppAction.PocketSponsoredStoriesChange -> {
```
I'm not familiar with how the Pocket code works. Does updating the setting necessitate updating the underlying data while on settings? I would have normally expected it to just be part of the Homescreen rendering process.

Mugurell replied:
The current flow assumes the stories composable just [observes the `pocketStories` state](https://github.com/mozilla-mobile/fenix/blob/64a12574b4a0ebd64cca91c0c9755f4928dc9972/app/src/main/java/org/mozilla/fenix/home/pocket/PocketStoriesViewHolder.kt#L56).
That is updated through `pocketStories = updatedStoriesState.getFilteredStories()` calls being done in the reducer.
If we emit `PocketSponsoredStoriesChange(emptyList())` from settings when the user disables the feature the reducer will update the AppState such that when the user goes back to homescreen and the composables read `pocketStories` it will only contain recommended stories.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/settings/HomeSettingsFragment.kt:
```diff
@@ -116,6 +126,44 @@ class HomeSettingsFragment : PreferenceFragmentCompat() {
+                            preference.isEnabled = false
+
+                            requireActivity().lifecycleScope.launch(Main) {
+                                val wasProfileDeletionSuccessful = withContext(ioDispatcher) {
+                                    context.components.core.pocketStoriesService.deleteProfile()
+                                }
```
I'm not familiar with the internals of Pocket. what's the TLDR of what's happening here/why it's necessary?

MozillaNoah reviewed (commented):
Ran it locally and it's looking good!

Mugurell commented:
Thank you Noah for the review!
Will address your suggestions and rebase after https://github.com/mozilla-mobile/fenix/pull/25353 is merged which adds similar changes to `Settings.kt` and `preference_keys.xml`.

Mugurell commented:
Rebased to include the recent merges and updated the behavior for when the feature is disabled based on https://mozilla-hub.atlassian.net/browse/FNXV2-20135?focusedCommentId=552310
